src/ripple/app/paths/impl/PaySteps.cpp

```diff
@@ -0,0 +1,533 @@
+                result.push_back (std::move (msr.second));
+                Currency dummy;
+                impliedPE.emplace (STPathElement::typeAccount,
+                    curIssue.account, dummy, curIssue.account);
+                cur = &*impliedPE;
+            }
+        }
+        else if (isOffer (cur) && isAccount (next))
+        {
+            if (curIssue.account != next->getAccountID () &&
+                !isXRP (next->getAccountID ()))
+            {
+                JLOG (j.trace) << "Inserting implied account after offer";
+                auto msr = make_DirectStepI (ctx(), curIssue.account,
+                    next->getAccountID (), curIssue.currency);
+                if (msr.first != tesSUCCESS)
+                    return {msr.first, Strand{}};
+                result.push_back (std::move (msr.second));
+            }
+            continue;
+        }
+
+        if (!isOffer (next) &&
+            hasCurrency(next) && next->getCurrency () != curIssue.currency)
+        {
+            auto const& nextCurrency = next->getCurrency ();
+            auto const& nextIssuer =
+                hasIssuer (next) ? next->getIssuerID () : curIssue.account;
+
+            if (isXRP (curIssue.currency))
+            {
+                JLOG (j.trace) << "Inserting implied XI offer";
+                auto msr = make_BookStepXI (
+                    ctx(), {nextCurrency, nextIssuer});
+                if (msr.first != tesSUCCESS)
+                    return {msr.first, Strand{}};
+                result.push_back (std::move (msr.second));
+            }
+            else if (isXRP (nextCurrency))
+            {
+                JLOG (j.trace) << "Inserting implied IX offer";
+                auto msr = make_BookStepIX (ctx(), curIssue);
+                if (msr.first != tesSUCCESS)
+                    return {msr.first, Strand{}};
+                result.push_back (std::move (msr.second));
+            }
+            else
+            {
+                JLOG (j.trace) << "Inserting implied II offer";
+                auto msr = make_BookStepII (
+                    ctx(), curIssue, {nextCurrency, nextIssuer});
+                if (msr.first != tesSUCCESS)
+                    return {msr.first, Strand{}};
+                result.push_back (std::move (msr.second));
+            }
+            impliedPE.emplace (
+                boost::none, nextCurrency, nextIssuer);
+            cur = &*impliedPE;
+            curIssue.currency = nextCurrency;
+            curIssue.account = nextIssuer;
+        }
+
+        auto s =
+            toStep (ctx (/*isLast*/ i == pes.size () - 2), cur, next, curIssue);
+        if (s.first == tesSUCCESS)
+            result.emplace_back (std::move (s.second));
+        else
+        {
+            JLOG (j.warning) << "toStep failed";
+            return {s.first, Strand{}};
+        }
+    }
+
+    return {tesSUCCESS, std::move (result)};
+}
+
+std::pair<TER, std::vector<Strand>>
+toStrands (
```

Comment:
Most often, I use lambdas for two things:

1) Complex initialization of a variable. This let's us replace code like:
```
Foo var;
{
   // Some code
   var = blockResult;
}
```
with:
```
auto const var = [&]
{
    // Some code
    return fooBlockResult;
}();
```

2) Local functions that require use of lots of the local function state. For example:
```
Foo(...a whole bunch of parameters...)
{
    // ...a whole bunch of local state
    auto helperFun = [&] {
        // ...use the parameters and local state
    }

    if (something)
    {
        helperFun();
    }

    while(something)
    {
        helperFun();
    }
}
```
If I made `helperFun` a free function I would have to pass a large number of parameters to it. The other alternative is a class to wrap local state and make `helperFun` a member var on that class.

Of these two cases, I think (1) is almost always OK. It's very easy to see what's going on. (2) is more of a problem. I should only use (2) when there are no other alternatives.

If I can refactor the code so the lambda doesn't need to capture so much local state, I can replace the lambda with a free function or small class.

Enough generalities, let's talk about this specific case.

What I'm doing with the `add` lambda is adding a strand to a `vector`, but I'm using the `vector` like a `set`. I create the strand, check if it's not already part of the set, and add it if it's not. I see three ways to improve this code:

1) Move the call `toStrand` outside the lambda. This means I no longer have to capture as many locals.

2) Write an `bool operator==(Strand const& lhs, Strand const& rhs)` free function. This removes one of the local lambdas and lets me use the `find` algorithm, which everyone knows.

3) Create a `StrandSet` class. This makes it clear what `add` is meant to do.

Note that I can't use something like `boost::flat_set` because I don't have `operator<` defined on `Step`. I did have an `operator<` defined at one point, but my implementation used `info` classes which was somewhat complex.

I also understand your comment wasn't necessarily about just this lambda. I'll be more thoughtful when using lambdas - especially type (2) lambdas.